storage/src/main/scala/ch/epfl/bluebrain/nexus/storage/files/CopyFiles.scala

```diff
@@ -0,0 +1,49 @@
+package ch.epfl.bluebrain.nexus.storage.files
+
+import cats.data.NonEmptyList
+import cats.effect.{IO, Ref}
+import cats.implicits.catsSyntaxParallelTraverse1
+import ch.epfl.bluebrain.nexus.storage.StorageError.CopyOperationFailed
+import fs2.io.file.{CopyFlag, CopyFlags, Files, Path}
+
+trait CopyFiles {
+  def copyValidated(name: String, files: NonEmptyList[ValidatedCopyFile]): IO[Unit]
+}
+
+object CopyFiles {
+  def mk(): CopyFiles = (name, files) =>
+    copyAll(name, files.map(v => CopyBetween(Path.fromNioPath(v.absSourcePath), Path.fromNioPath(v.absDestPath))))
+
+  final private[files] case class CopyBetween(source: Path, dest: Path)
+
+  def copyAll(name: String, files: NonEmptyList[CopyBetween]): IO[Unit] =
```

Comment:
This shouldn't be public, the test can just use an instance

Reply:
Also I don't know if it's necessarily helpful to check the list is nonempty here. I think practically speaking nothing bad will happen if an empty list is passed here. Sometimes it's easier in calling code if you just allow it and do nothing

Reply:
The problem here was that `ValidatedCopyFile` is intentionally hard to construct, and tests for copying shouldn't need to care about the validation logic.

On the NEL point I just tend to prefer a stricter data structure when possible. It communicates more information. Once the parsing into a stricter structure is done at the edges then assumptions we have about that data are explicit in the type rather than implicit. In this case I'd rather reject the request in the endpoint for example

Reply:
I generally agree, it was also because I had my eye on reusing this code in delta so it will be moved to the kernel. The answer is probably to change the abstraction to use that function. I'll do that when I move it in the delta change ?